File: java/src/main/java/fr/arolla/Inventory.java
package fr.arolla;

public class Inventory implements Listener {
  private final EventBus bus;
  private int capacity;

  public Inventory(EventBus bus, int totalSeats) {
    this.bus = bus;
    this.bus.subscribe(this);
    capacity = totalSeats;
  }

  public void decrementCapacity(int numSeats) {
    if (numSeats > capacity) {
      throw new OverBookedException();
    }

    capacity -= numSeats;
    System.out.format("Capacity now at %d\n", capacity);
  }

  public int getCapacity() {
    return capacity;
  }

  @Override
  public void onMessage(Object message) {
    if (message instanceof BookingRequested bookingRequested) {
      var numSeats = bookingRequested.numSeats();
      var user = bookingRequested.user();
      try {
        decrementCapacity(numSeats);
        bus.emit(new CapacityUpdated(capacity, numSeats, user));
      } catch (OverBookedException e) {
        bus.emit(new CapacityExceeded(user));
      }
    }
  }
}

File: dotnet/Choreography/Inventory.cs
using Choreography.Events;

namespace Choreography;

public class Inventory : ISubscriber
{
    private readonly Context _context;
    private int _capacity;


    public Inventory(Context context, int totalSeats)
    {
        _capacity = totalSeats;
        _context = context;
        _context.Bus().Subscribe(this);
    }

    public void OnEvent(IEvent e)
    {
        var bus = _context.Bus();
        if (e is BookingRequested bookingRequested)
        {
            var numSeats = bookingRequested.NumSeats;
            try
            {
                DecrementCapacity(numSeats);

                bus.Emit(new CapacityUpdated(_capacity, numSeats));
            }
            catch (OverbookedException)
            {
                bus.Emit(new OverBooked());
            }
        }
    }

    public int CurrentCapacity()
    {
        return _capacity;
    }

    public void DecrementCapacity(int numSeats)
    {
        if (numSeats > _capacity)
        {
            _context.Logger().Log("OverBooked");
            throw new OverbookedException();
        }

        _capacity -= numSeats;
        _context.Logger().Log($"Capacity is now at {_capacity}");
    }
}